### PR TITLE
Harden Operator CI

### LINKED ISSUE #10587: Flaky KeycloakServicesE2EIT.testDiscoveryServiceDurability
### Describe the bug

The operator test: `KeycloakServicesE2EIT.testDiscoveryServiceDurability` spuriously failed here:

https://github.com/keycloak/keycloak/runs/5413342790?check_suite_focus=true

### Version

latest

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by andreaTP**
Hopefully 🤞  resolves #10587

Still running some jobs for additional checks, opening it anyhow to speed up the review process.

### CODE REVIEW

File: .github/workflows/operator-ci.yml
Comment (andreaTP) [inline .github/workflows/operator-ci.yml:17]: I'll remove those lines as soon as the CI completes.

**Comment by andreaTP** [comment]
20 CI runs just passed here:
https://github.com/andreaTP/keycloak/actions/runs/1952332120

File: .github/workflows/operator-ci.yml
Comment (andreaTP) [reply .github/workflows/operator-ci.yml]: removed

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java
Comment (andreaTP) [inline operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java:50]: This is the key change, disabling `generationAwareEventProcessing` workaround the race condition we found.

Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java:50]: Can you please create a follow-up task with all the linked issues that need to be resolved to unblock it?

File: operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java
Comment (vmuzikar) [inline operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java:95]: Not sure why to make this change but ok. 😉

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java
Comment (vmuzikar) [inline operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java:117]: I presume we need these ugly `if`s because of `generationAwareEventProcessing = false`. Correct?
So we're effectively managing the reconciliation loop and relying on time checks instead of events? That means the CRs are not watched any more?

If all the answers are "yes", I'd rather see `rescheduleAfter` even when `ready == true`. What if the deployment fails suddenly? The operator would not know.

**Review by vmuzikar — CHANGES_REQUESTED** [review]

**Comment by andreaTP** [comment]
Quick update, with ~ those changes 200 runs are green:
https://github.com/andreaTP/keycloak/actions/runs/1952340354

File: operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java
Comment (andreaTP) [reply operator/src/test/java/org/keycloak/operator/utils/K8sUtils.java:95]: Just made imports uniform

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java:117]: > I presume we need these ugly ifs because of `generationAwareEventProcessing = false`. Correct?

No, this is just hardening the logic for retrying the reconciliation.

> So we're effectively managing the reconciliation loop and relying on time checks instead of events? That means the CRs are not watched any more?

No, this change simply means: "if we know that there is additional processing to be done from the operator we install a fail-safe temporal fallback".
You can read more about [here](https://javaoperatorsdk.io/docs/features) in the section "Retry and Rescheduling and Event Handling Common Behavior"

> What if the deployment fails suddenly? The operator would not know.

We should keep relying on events, and we can possibly remove those safe-nets later on in the development.

**Comment by andreaTP** [comment]
@vmuzikar :

> Can you please create a follow-up task with all the linked issues that need to be resolved to unblock it?

Sure I'll create an issue tomorrow as first thing.

Ready for a second review round. I hope we can get this merged asap to not cause friction and noise in the rest of the PRs.

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java
Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java:117]: Thanks for the clarification. In that case, could we do something like this?

```java
UpdateControl<Keycloak> ret;

        if (status.equals(kc.getStatus())) {
            ret = UpdateControl.noUpdate();
        }
        else {
            ret = UpdateControl.updateStatus(kc);
        }
        
        if (status.getConditions().stream()
                .anyMatch(c -> c.getType().equals(KeycloakStatusCondition.READY) && !c.getStatus())) {
            ret.rescheduleAfter(10, TimeUnit.SECONDS);
        }

        return ret;
```

I hate that we are now introducing additional logic for detecting ready state while we already have this information.

Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java]: Unfortunately the proposed change invalidates the fail over logic:
```java
status.getConditions().stream()
        .anyMatch(c -> c.getType().equals(KeycloakStatusCondition.READY)
```
Won't detect the fact that a [`service` needs to be re-created](https://github.com/keycloak/keycloak/pull/10653/files#diff-6a6cb293bff38002e801b4adaf350b0bb8ead4599ad52903b654fde6ddcb4350R88).

I anyhow refactored the logic to be a little nicer to the eyes.

@vmuzikar those fail over logics and the workarounds needs to be audited and reviewed as soon as we have a proper fix from the JOSDK.

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java:50]: Follow up issue #10658

**Comment by andreaTP** [comment]
Follow up issue: #10658
Upstream issue: https://github.com/java-operator-sdk/java-operator-sdk/issues/1006

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java
Comment (andreaTP) [reply operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakController.java]: this should be resolved now @vmuzikar

File: operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java
Comment (vmuzikar) [inline operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java:513]: This is an unnecessary change, but ok, not blocking.

**Review by vmuzikar — APPROVED** [review]

**Review by abstractj — APPROVED** [review]

**Comment by andreaTP** [comment]
@abstractj can you merge this please?

**Pull request merged**
